### DISCUSSION TITLE
Autofilling on `/#/settings/security/change-password` will expose password

### DISCUSSION THREAD

**Discussion by bilogic** [dani-garcia/vaultwarden#3937] [Q&A]
While trying to test out multi user behavior, I saved the various user passwords into my vault.

When loading `/#/settings/security/change-password`, the Chrome extension autofills the `Current master password` and `Master password hint`. The problem is, `Master password hint` is in plain text and it exposes the autofilled password.

Is there a way to break this behavior? Preferrably on just this page, perhaps by renaming the `input` tag's `id` that is `masterPasswordHint` to something else?

### Your environment (Generated via diagnostics page)
* Vaultwarden version: v1.29.2
* Web-vault version: v2023.7.1
* OS/Arch: linux/x86_64
* Running within Docker: true (Base: Debian)
* Environment settings overridden: false
* Uses a reverse proxy: true
* IP Header check: true (X-Real-IP)
* Internet access: true
* Internet access via a proxy: false
* DNS Check: true
* Browser/Server Time Check: true
* Server/NTP Time Check: true
* Domain Configuration Check: true
* HTTPS Check: true
* Database type: SQLite
* Database version: 3.41.2
* Clients used: 
* Reverse proxy and version: 
* Other relevant information: 

### Config (Generated via diagnostics page)
<details><summary>Show Running Config</summary>

**Environment settings which are overridden:** 

```json
{
  "_duo_akey": null,
  "_enable_duo": false,
  "_enable_email_2fa": false,
  "_enable_smtp": true,
  "_enable_yubico": false,
  "_icon_service_csp": "",
  "_icon_service_url": "",
  "_ip_header_enabled": true,
  "_smtp_img_src": "cid:",
  "admin_ratelimit_max_burst": 3,
  "admin_ratelimit_seconds": 300,
  "admin_session_lifetime": 20,
  "admin_token": "***",
  "allowed_iframe_ancestors": "",
  "attachments_folder": "/data/attachments",
  "auth_request_purge_schedule": "30 * * * * *",
  "authenticator_disable_time_drift": false,
  "data_folder": "/data",
  "database_conn_init": "",
  "database_max_conns": 10,
  "database_timeout": 30,
  "database_url": "****************",
  "db_connection_retries": 15,
  "disable_2fa_remember": false,
  "disable_admin_token": false,
  "disable_icon_download": false,
  "domain": "*****://***********",
  "domain_origin": "*****://***********",
  "domain_path": "",
  "domain_set": true,
  "duo_host": null,
  "duo_ikey": null,
  "duo_skey": null,
  "email_attempts_limit": 3,
  "email_expiration_time": 600,
  "email_token_size": 6,
  "emergency_access_allowed": true,
  "emergency_notification_reminder_schedule": "0 3 * * * *",
  "emergency_request_timeout_schedule": "0 7 * * * *",
  "enable_db_wal": true,
  "event_cleanup_schedule": "0 10 0 * * *",
  "events_days_retain": null,
  "extended_logging": true,
  "helo_name": null,
  "hibp_api_key": null,
  "icon_blacklist_non_global_ips": true,
  "icon_blacklist_regex": null,
  "icon_cache_folder": "/data/icon_cache",
  "icon_cache_negttl": 259200,
  "icon_cache_ttl": 2592000,
  "icon_download_timeout": 10,
  "icon_redirect_code": 302,
  "icon_service": "internal",
  "incomplete_2fa_schedule": "30 * * * * *",
  "incomplete_2fa_time_limit": 3,
  "invitation_expiration_hours": 120,
  "invitation_org_name": "Avexsoft",
  "invitations_allowed": true,
  "ip_header": "X-Real-IP",
  "job_poll_interval_ms": 30000,
  "log_file": null,
  "log_level": "Info",
  "log_timestamp_format": "%Y-%m-%d %H:%M:%S.%3f",
  "login_ratelimit_max_burst": 10,
  "login_ratelimit_seconds": 60,
  "org_attachment_limit": null,
  "org_creation_users": "",
  "org_events_enabled": false,
  "org_groups_enabled": false,
  "password_hints_allowed": true,
  "password_iterations": 100000,
  "push_enabled": false,
  "push_installation_id": "***",
  "push_installation_key": "***",
  "push_relay_uri": "https://push.bitwarden.com",
  "reload_templates": false,
  "require_device_email": false,
  "rsa_key_filename": "/data/rsa_key",
  "send_purge_schedule": "0 5 * * * *",
  "sendmail_command": null,
  "sends_allowed": true,
  "sends_folder": "/data/sends",
  "show_password_hint": false,
  "signups_allowed": false,
  "signups_domains_whitelist": "",
  "signups_verify": true,
  "signups_verify_resend_limit": 6,
  "signups_verify_resend_time": 3600,
  "smtp_accept_invalid_certs": false,
  "smtp_accept_invalid_hostnames": false,
  "smtp_auth_mechanism": null,
  "smtp_debug": false,
  "smtp_embed_images": true,
  "smtp_explicit_tls": null,
  "smtp_from": "**********************",
  "smtp_from_name": "Vaultwarden",
  "smtp_host": "**************",
  "smtp_password": "***",
  "smtp_port": 587,
  "smtp_security": "starttls",
  "smtp_ssl": null,
  "smtp_timeout": 15,
  "smtp_username": "******************",
  "templates_folder": "/data/templates",
  "tmp_folder": "/data/tmp",
  "trash_auto_delete_days": null,
  "trash_purge_schedule": "0 5 0 * * *",
  "use_sendmail": false,
  "use_syslog": false,
  "user_attachment_limit": null,
  "web_vault_enabled": true,
  "web_vault_folder": "web-vault/",
  "websocket_address": "0.0.0.0",
  "websocket_enabled": true,
  "websocket_port": 8080,
  "yubico_client_id": "******",
  "yubico_secret_key": "***",
  "yubico_server": null
}
```
</details>

**Comment by BlackDex** [comment]
So, this is a client behavior item. Vaultwarden doesn't maintain the clients. Also, if you don't want the password hint to be shown, i would suggest to rename that field in the item which gets auto-filled, like add spaces between the words or `-` or `_` for example.

**Reply by BlackDex** [reply]
Also, changing the ID on the html side is probably not a good idea, ot probably is used by several items with that name.

**Reply by bilogic** [reply]
![image](https://github.com/dani-garcia/vaultwarden/assets/946010/2840d515-d4cc-4e9f-ae03-a5a461e1eb7d)

Hang on, the web UI is not maintained by Vaultwarden? Why do I see the VW logo top left corner?
So which repository controls this page then? https://selfhostedvaultwarden/#/settings/security/change-password

The `Master password hint` field exposes the autofilled password.

**Reply by BlackDex** [reply]
We make soms small modifications like change the logo and names to prevent confusions between Bitwarden and Vaultwarden so that people do not instantly report any issue to Bitwarden, which seems to help 😉. The web-vault patched here https://github.com/dani-garcia/bw_web_builds

The original client code can be found here https://github.com/bitwarden/clients

If you can mimic this on vault.bitwarden.com, which I think you can, then i suggest to report an issue over there, or maybe first look for reported issues there mentioning the same.

**Reply by kpfleming** [reply]
This could only be happening if you have your password for BW/VW itself *in* your vault, right? That doesn't seem like a normal thing to do.

**Reply by bilogic** [reply]
@BlackDex thanks!

@kpfleming no it's not, but I think it is still something that should be fixed for at least 2 reasons:
- the Chrome extension is not interpreting fields properly (one should never autofill a plaintext field)
- webUI and Chrome extension is not playing nicely together

**Reply by BlackDex** [reply]
I have use cases for plain text auto fill. Also, usernames are most of the time plain text fields.

Also, Bitwarden doesn't autofill site A items on site B, so you must have stored something which has your Vaultwarden instance domain there.

**Reply by bilogic** [reply]
@BlackDex I meant filling a password into a plain text field, I'm aware of the domain checking before filling. Again, I was testing multi user behavior in BW/VW. To make it less tedious for myself, I stored their credentials into the vault.

Anyway, the proposal I have is rename `masterPasswordHint` to something without the word `Password`, that will probably "fix" this behavior without affecting your use cases or much else

**Reply by BlackDex** [reply]
Did you marked the field as hidden? Not sure that it would matter, since there are probably always use cases where some site uses a plaintext field and you want it hidden. So, in my opinion, i would rename the stored field of the item, which solves your case.

**Reply by bilogic** [reply]
Could explain marking the field as hidden? And what does rename stored field mean too?

My entry stores `username` + `password`. I wouldn't rename `password` since I have to login to BW/VW's webUI login screen.
Or did you mean I can hide the password hint field? 

But I'm not about to modify any source code in my hosted VW, that would really be opening a can of worms.

**Reply by stefan0xC** [reply]
You could add a new custom field to your entry named `masterPasswordHint` with your password hint (or a dummy text if you don't have one) so the client will fill the field with something else besides your password...

**Reply by bilogic** [reply]
@stefan0xC thanks that helped, and BW side accepted my report. After some thought, I think this is an unintended consequence by both user and dev, so I still think something should be done about it; since seeing my own password in plaintext can come as a shock. I'm looking forward to seeing what BW will do about it.
